tfx/experimental/pipeline_testing/pipeline_recorder_utils.py

```diff
@@ -0,0 +1,151 @@
+
+from collections import defaultdict
+import os
+
+import absl
+from ml_metadata.proto import metadata_store_pb2
+import tensorflow as tf
+from typing import Dict, List, Text
+
+from tfx.orchestration import metadata
+from tfx.utils import io_utils
+
+def get_paths(metadata_connection: metadata.Metadata,
+              executions: Dict[Text, List[metadata_store_pb2.Execution]],
+              output_dir: Text):
+  """Returns a zipped list of source artifact uris and destination uris.
+  Destination uris are stored in output_dir.
+  Args:
+    metadata_connection: A class for metadata I/O to metadata db.
+    executions:
+    output_dir: Uri to metadata db.
+
+  Returns:
+    Zip of src_uris and dest_uris.
+  """
+  events = [
+      x for x in metadata_connection.store.get_events_by_execution_ids(
+          [e.id for e in executions])
+      if x.type == metadata_store_pb2.Event.OUTPUT
+  ]
+  unique_artifact_ids = list({x.artifact_id for x in events})
+
+  src_uris = []
+  dest_uris = []
+  for artifact in \
+        metadata_connection.store.get_artifacts_by_id(unique_artifact_ids):
+    src_uris.append(artifact.uri)
+    component_id = \
+        artifact.custom_properties['producer_component'].string_value
+    name = artifact.custom_properties['name'].string_value
+    dest_uris.append(os.path.join(output_dir, component_id, name))
+  return zip(src_uris, dest_uris)
+
+def get_execution_dict(metadata_connection: metadata.Metadata
+                      ) -> Dict[Text, List[metadata_store_pb2.Execution]]:
+  """Returns dictionary mapping holding executions for run_id.
+
+  Args:
+    metadata_connection: A class for metadata I/O to metadata db.
+
+  Returns:
+    A dictionary that holds executions for pipeline run_id
+  """
+  execution_dict = defaultdict(list)
+  for execution in metadata_connection.store.get_executions():
+    execution_run_id = execution.properties['run_id'].string_value
+    execution_dict[execution_run_id].append(execution)
+  return execution_dict
+
+def get_latest_executions(metadata_connection: metadata.Metadata,
+                          pipeline_name: Text
+                          ) -> List[metadata_store_pb2.Execution]:
+  """Fetches executions associated with the latest context.
+  Args:
+    metadata_connection: A class for metadata I/O to metadata db.
+
+  Returns:
+    A dictionary that holds executions for pipeline run_id
+  """
+  latest_context_id = None
+  latest_updated_time = float('-inf')
+
+  for pipeline_run_context in \
+      metadata_connection.store.get_contexts_by_type(
+          metadata._CONTEXT_TYPE_PIPELINE_RUN):  # pylint: disable=protected-access
+    if pipeline_name != \
+          pipeline_run_context.properties['pipeline_name'].string_value:
+      continue
+    if latest_updated_time < pipeline_run_context.last_update_time_since_epoch:
+      latest_updated_time = pipeline_run_context.last_update_time_since_epoch
+      latest_context_id = pipeline_run_context.id
```

Comment:
It would be simpler if we use `max()`, for example,
```
pipeline_run_contexts = metadat_connection.store.get_...()
latest_context = max(pipeline_run_contexts, key=lambda c: c.last_update_time_since_epoch)
```